### PR TITLE
[SPARK-47380][CONNECT] Ensure on the server side that the SparkSession is the same

### PR DESCRIPTION
**Pull request by nemanja-boric-databricks**
### What changes were proposed in this pull request?

In this PR we change the client behaviour to send the previously observed server session id so that the server can validate that the client used to talk with this specific session. Previously this was only validated on the client side which made the server actually execute the request for the wrong session before throwing on the client side (once the response from the server was obtained).

### Why are the changes needed?
The server can execute the client command on the wrong spark session before client figuring out it's the different session.

### Does this PR introduce _any_ user-facing change?
The error message now pops up differently (it used to be a slightly different message when validated on the client).

### How was this patch tested?
Existing unit tests, add new unit test, e2e test added, manual testing

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by nemanja-boric-databricks** [comment]
Failed suite is:

```
[info] *** 1 SUITE ABORTED ***
[error] Error during tests:
[error] 	org.apache.spark.sql.jdbc.v2.OracleIntegrationSuite
```

But it passed for the previous commits, it seems unrelated.

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala [hunk 1]
```diff
@@ -252,8 +268,7 @@ private[sql] class SparkConnectClient(
       .setUserContext(userContext)
       .setSessionId(sessionId)
       .setClientType(userAgent)
-      .build()
-    bstub.releaseSession(request)
+    bstub.releaseSession(request.build())
```
Comment (hvanhovell) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala:271]: These two lines are equivalent right?

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by hvanhovell** [comment]
Merging test failure is unrelated.

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala (hunk 1 above)
Comment (nemanja-boric-databricks) [reply connector/connect/common/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala:271]: Yes, this is just to make it uniform to everything else.

**Pull request closed without merging**